### PR TITLE
feat: add async transaction request interface

### PR DESCRIPTION
**Pull request by sulenn**

### CODE REVIEW

**Comment by fisco-dev** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- conn/channel.go  5
         

Complexity decreasing per file
==============================
+ abi/bind/base.go  -1
         

Clones added
============
- abi/bind/base.go  2
- conn/channel.go  2
- precompiled/config/system_config.go  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/FISCO-BCOS/go-sdk/pullRequest?prid=6066259&bid=19867145)

File: abi/bind/base.go [hunk 1]
```diff
@@ -118,6 +118,21 @@ func DeployContract(opts *TransactOpts, abi abi.ABI, bytecode []byte, backend Co
 	return c.address, tx, c, nil
 }
 
+func AsyncDeployContract(handler func(*types.Receipt, error), opts *TransactOpts, abi abi.ABI, bytecode []byte, backend ContractBackend, params ...interface{}) (*types.Transaction, *BoundContract, error) {
```
Comment (bxq2011hust) [inline abi/bind/base.go:121]:
```suggestion
func AsyncDeployContract(opts *TransactOpts, handler func(*types.Receipt, error), abi abi.ABI, bytecode []byte, backend ContractBackend, params ...interface{}) (*types.Transaction, *BoundContract, error) {
```

File: conn/channel.go [hunk 2]
```diff
@@ -501,13 +503,11 @@ func (hc *channelSession) sendTransaction(ctx context.Context, msg interface{})
 		return nil, fmt.Errorf("send transaction error, code=%d, message=%s", respmsg.Error.Code, respmsg.Error.Message)
 	}
 	// fmt.Printf("sendTransaction reveived response,seq:%s message:%s\n ", rpcMsg.uuid, respmsg.Result)
-
 	<-receiptResponse.Notify
-
-	hc.mu.RLock()
+	hc.mu.Lock()
```
Comment (bxq2011hust) [inline conn/channel.go:507]: Read lock is enough？

**Comment by fisco-dev** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- conn/channel.go  5
         

Complexity decreasing per file
==============================
+ abi/bind/base.go  -1
         

Clones added
============
- conn/channel.go  2
- precompiled/config/system_config.go  2
- abi/bind/base.go  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/FISCO-BCOS/go-sdk/pullRequest?prid=6066259&bid=19867145)

File: conn/connection.go [hunk 3]
```diff
@@ -307,6 +308,21 @@ func (c *Connection) CallContext(ctx context.Context, result interface{}, method
 	}
 }
 
+func (c *Connection) AsyncCallContext(ctx context.Context, handler func(*types.Receipt, error), method string, args ...interface{}) error {
```
Comment (bxq2011hust) [inline conn/connection.go:311]:
```suggestion
func (c *Connection) AsyncSendTransaction（ctx context.Context, handler func(*types.Receipt, error), method string, args ...interface{}) error {
```

File: client/api_handler.go [hunk 4]
```diff
@@ -151,6 +150,45 @@ func (api *APIHandler) SendRawTransaction(ctx context.Context, groupID int, tx *
...
+					return
+				}
+				if err != nil {
+					errorStr := fmt.Sprintf("%s", err)
+					if strings.Contains(errorStr, "connection refused") {
+						handler(nil, errors.New("connection refused"))
+						return
+					}
+				} else {
+					fmt.Println("Transaction not yet mined")
```
Comment (bxq2011hust) [inline client/api_handler.go:179]:
```suggestion
					//fmt.Println("Transaction not yet mined")
```

File: conn/channel.go [hunk 5]
```diff
@@ -1017,10 +1074,34 @@ func (hc *channelSession) processMessages() {
...
-				} else {
-					fmt.Printf("error %+v", receipt)
 				}
 				hc.mu.Unlock()
+				hc.asyncMu.Lock()
+				if handler, ok := hc.asyncHandlers[msg.uuid]; ok {
+					if msg.errorCode != 0 {
+						handler(nil, errors.New("response error:"+string(msg.errorCode)))
+					} else {
+						var receipt = new(types.Receipt)
```
Comment (bxq2011hust) [inline conn/channel.go:1084]: can we parse Receipt use a func

**Comment by fisco-dev** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- conn/channel.go  5
         

Complexity decreasing per file
==============================
+ abi/bind/base.go  -1
         

Clones added
============
- conn/channel.go  2
- abi/bind/base.go  2
- precompiled/config/system_config.go  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/FISCO-BCOS/go-sdk/pullRequest?prid=6066259&bid=19867145)

**Review by bxq2011hust — APPROVED** [review]

**Pull request merged**